Analysis Helper Classes / Functions

lhwhitehead opened the pull request:
My first attempt at a pull request, so I hope it works ok! This contains the first efforts at adding the analysis utility classes. There are no doubt many other useful functions that we can add in the future... but these are the initial ones that I have implemented.

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraUtilsBase.h:
```diff
@@ -0,0 +1,90 @@
+ * @brief LArPandoraUtilsBase class containing some template functions
+ *
+*/
+class LArPandoraUtilsBase
```
As you pointed out, in other repositories(e.g. LArClusterHelper in larcontent) it looks like the standard is to group related functions in a class and define them public static.  It looks like consistent style is important across the repos which would mean (potentially) a large number of changes to swap to free standing functions grouped in namespace...

lhwhitehead replied:
Dom and I were discussing this and the advantage we thought of for the class approach is that no one can add other functions in other code blocks, unlike the namespace method.

PandoraPFA replied:
I agree about Andy's "not strictly OO" point, but this is consistent with other areas of Pandora, where this has been done for the reasons you suggest. So fine to note and leave as is.

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraUtilsBase.h:
```diff
@@ -0,0 +1,90 @@
+};
+
+    // Implementation of the template function to get the products from the event
+    template <typename T> void LArPandoraUtilsBase::GetProductVector(art::Event const &evt, const std::string &label, std::vector<art::Ptr<T>> &productVector)
```
Do you mean moving this to a usual .cc file or something like a .tcc file that gets included in the header?  For the former, wouldn't Leigh have to define each input object case for the template implementation?

loressa replied:
.tcc? 

Maybe this serves as example: 

https://github.com/loressa/LArContent/blob/master/larpandoracontent/LArVertex/MvaVertexSelectionAlgorithm.cc#L902

Compare to the templated implementation in the header file:
https://github.com/loressa/LArContent/blob/master/larpandoracontent/LArVertex/MvaVertexSelectionAlgorithm.h

absolution1 replied:
Right.  In the case Lorena has pointed out, that template is only defined for two input objects whereas Leigh's template needs to work for every possible input type, right?

AndyChappell replied:
Fair point, this solution would be more relevant to the namespace-based implementation noted below, rather than the class hierarchy solution - as then the implementation would still be accessible to the other functions, as they'd all reside in the same source file. In principle you could include the LArPandoraUtilsBase.cc in the inheriting class source files, but that's not really what I had in mind.

lhwhitehead replied:
I disagree with this one. Moving the code from the header means that every time someone thinks of a new product in art that we could access then they have to add a line to the implantation file. It is totally generic and extendable without changes the way the code is written now.

absolution1 replied:
Yes.  As I mentioned above, if you wanted to make the header file smaller in number of lines, you could move the contents to a completely separate file that you #include at the bottom of the header, but I think that's mostly a style thing.

PandoraPFA replied:
It's not the end of the world to have to enumerate the different types that you are wanting to fit the requirements placed on the `t` variables of type `T` in your compiler-generated implementations. I agree that there's a lot of precedent (and some small precedent in Pandora) for leaving the template implementation in the header file. Then it's possible to create a new version of the function (for a new type) at compile time for a new package that includes the header file. If the implementation was in your cc file, you obviously need to tell compiler which versions should be exported to the library when you built your original package.

Dom's idea keeps header file cleaner, but isn't something I've seen used commonly. Preference to keep header file quite clean by leaving the implementation outside of the class declarations, then inlining it below. i.e. class declarations shouldn't be affected in any way by implementation choices.

Leigh's call, after this nice discussion of all of the relevant issues we could think of.

lhwhitehead replied:
So I've left the implementation in the head file, but I've improved the way the template functions work, so now most utility functions can just return the result of the templates in the base class. I think this wrapping of the template functions is easier for users to deal with compared to typedefs of the functions (though it might feel like the best coding solution).

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraUtilsBase.h:
```diff
@@ -0,0 +1,90 @@
+    {
+
+        art::Handle<std::vector<T>> theseProds;
+        evt.getByLabel(label,theseProds);
```
I *believe* that getByLabel returns a success bool so you could contract the next few lines slightly

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraUtilsBase.h:
```diff
@@ -0,0 +1,90 @@
+        if (!theseProds.isValid())
+        {
+            mf::LogError("LArPandora") << " Failed to find product with label " << label << " ... returning empty vector" << std::endl;
+            productVector = std::vector<art::Ptr<T>>();
```
Pretty sure we discussed this already but do art::Ptr valid checks do sensible things for a default-constructed art::Ptr?

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraUtilsBase.h:
```diff
@@ -0,0 +1,90 @@
+
+        // We need to convert these to art pointers
+        productVector.clear();
+        for(unsigned int i = 0; i < theseProds->size(); ++i){
```
I do this too but size_t saves button presses

PandoraPFA replied:
Be wary about portability though!

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraUtilsBase.h:
```diff
@@ -0,0 +1,90 @@
+    {
+
+        art::Handle<std::vector<U>> products;
+        evt.getByLabel(label,products);
```
As above, I think this function returns a success bool

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraUtilsBase.h:
```diff
@@ -0,0 +1,90 @@
+
+        const std::vector<art::Ptr<T>> theseAssocs = findParticleAssocs.at(prod.key());
+
+        productVector = theseAssocs;
```
Do you need the assignment here?  Could the above line be

`productVector = findParticleAssocs.at(prod.key());`

lhwhitehead replied:
Done!

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraTrackUtils.h:
```diff
@@ -0,0 +1,77 @@
+
+#include "art/Framework/Principal/Event.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "canvas/Persistency/Common/FindManyP.h"
```
includes in header and implementation

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraTrackUtils.cxx:
```diff
@@ -0,0 +1,53 @@
+        return theseSPs;
+    }
+
+    const art::Ptr<recob::PFParticle> LArPandoraTrackUtils::GetParticle(const art::Ptr<recob::Track> &track, art::Event const &evt, const std::string &label)
```
Function name should probably be PFParticle to avoid the ambiguity

lhwhitehead replied:
Agreed

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraTrackUtils.h:
```diff
@@ -0,0 +1,77 @@
+    * 
+    * @return art::Ptr to the particle
+    */
+    static const art::Ptr<recob::PFParticle> GetParticle(const art::Ptr<recob::Track> &track, art::Event const &evt, const std::string &label);
```
Function name should probably be PFParticle to avoid the ambiguity

lhwhitehead replied:
Agreed

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraShowerUtils.h:
```diff
@@ -0,0 +1,77 @@
+    * 
+    * @return art::Ptr to the particle 
+    */
+    static const art::Ptr<recob::PFParticle> GetParticle(const art::Ptr<recob::Shower> &shower, art::Event const &evt, const std::string &label);
```
GetParticle -> GetPFParticle avoids ambiguity

lhwhitehead replied:
Agreed

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraShowerUtils.h:
```diff
@@ -0,0 +1,77 @@
+    */
+    static const art::Ptr<recob::PFParticle> GetParticle(const art::Ptr<recob::Shower> &shower, art::Event const &evt, const std::string &label);
+
+private:
```
Could this be dropped?

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.h:
```diff
@@ -0,0 +1,100 @@
+    * 
+    * @return vector of art::Ptrs to the child particles 
+    */
+    static const std::vector<art::Ptr<recob::PFParticle>> GetChildParticles(const art::Ptr<recob::PFParticle> &part, art::Event const &evt, const std::string &label);
```
Particle -> PFParticle avoid ambiguity

PandoraPFA replied:
Agree.

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+    {
+
+        art::Handle<std::vector<recob::PFParticle>> particles;
+        evt.getByLabel(label,particles);
```
There should be a handle validity check as this could explode.

lhwhitehead replied:
Good find

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+
+    const std::vector<art::Ptr<recob::SpacePoint>> LArPandoraPFParticleUtils::GetSpacePoints(const art::Ptr<recob::PFParticle> &part, art::Event const &evt, const std::string &label)
+    {
+
```
accidental empty line?

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+namespace lar_pandora
+{
+
+    const std::vector<art::Ptr<anab::T0>> LArPandoraPFParticleUtils::GetT0(const art::Ptr<recob::PFParticle> &part, art::Event const &evt, const std::string &label)
```
Mostly just me musing.
A lot of these functions do exactly the same thing so there's probably room for template'ing
But, these functions are helper functions and so should be as easy to use as possible.  I'm probably in the minority here but I think putting templates in places opens up the possibility of confusing the end user.
So maybe the boilerplating necessary here.

Bit of a pointless comment really...

PandoraPFA replied:
I think templating hidden implementations is a good idea, then you can use some nice typedefs, if you wish, to make things nice and user-friendly. It's all about the user experience.

lhwhitehead replied:
I was wondering about something like that when writing this (repetitive) code... I guess I could template them and have the public methods as typedefs of the templated function?

PandoraPFA replied:
Should be easier to maintain, so (if you have a chance) give it a go and see whether you like it or not, then it's your call :)

lhwhitehead replied:
I've thought about this a lot, and I think I have a middle-ground solution. I have improved the templating in the base class to have return types instead of passing by argument. So now the individual functions for the different helpers are one-liners, but I think it keeps improved readability for users than just using typedefs.

lhwhitehead replied:
some pseudo code:
`std::vector<art::Ptr<recob::Hit>> GetHits(art::Ptr<recob::Track> &track)
{
    return GetAssocProductVector<recob::Hit>GetHits(track)
}
`

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+    {
+        // This function needs to fail if GetTrack would fail
+        std::vector<art::Ptr<recob::Track>> theseTracks;
+        GetAssocProductVector(part,evt,particleLabel,trackLabel,theseTracks);
```
Was getting the associated track a deliberate choice here?
For analysis stuff I've always checked the PDG of the PFP.
I guess if more PDGs are used in the future the implementation I've described might break.

My only other thought about this is end users doing something like

`
if (IsTrack(...){
  art::Ptr<Track> track = GetTrack(...);
}
`
the same functionality under the hood would be called twice.

This method is more robust than the way I described above so maybe the (most likely extremely tiny) efficiency loss is worthwhile....

lhwhitehead replied:
The PDG approach has the failing that we sometimes get no track reconstructed for a "track-like" PDG. Ideally I would use the approach that you outlined (and indeed did on my first attempt), but unfortunately the reality doesn't let me.

absolution1 commented on larpandora/LArPandoraAnalysis/UtilityExample_module.cc:
```diff
@@ -0,0 +1,187 @@
+    std::cout << "Found the neutrino, and it has pdg code " << neutrino->PdgCode() << " and " << neutrinoChildren.size() << " child particles" << std::endl;
+
+    // Lets see how many of the children are track- or shower-like
+    for (unsigned int c = 0; c < neutrinoChildren.size(); ++c)
```
I usually opt for good ol' indices but I'm not against using iterators.  but, as this is an example module, I'd personally opt for indices.

absolution1 replied:
Although, if my fingers remembered, I'd use size_t.
Also changing the index 'c' to 'iChild' makes it a bit clearer.

absolution1 commented on larpandora/LArPandoraAnalysis/UtilityExample_module.cc:
```diff
@@ -0,0 +1,187 @@
+
+void UtilityExample::reconfigure(fhicl::ParameterSet const &pset)
+{
+    m_particleLabel = pset.get<std::string>("PFParticleModule","pandora");
```
These should be constructed in the constructor list.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraClusterUtils.h:
```diff
@@ -0,0 +1,55 @@
+    */
+    static const std::vector<art::Ptr<recob::Hit>> GetHits(const art::Ptr<recob::Cluster> &cluster, art::Event const &evt, const std::string &label);
+
+private:
```
I would remove this (and other such examples) as redundant, and trim stray whitespace.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraEventUtils.cxx:
```diff
@@ -0,0 +1,150 @@
+namespace lar_pandora
+{
+
+    const std::vector<art::Ptr<recob::PFParticle>> LArPandoraEventUtils::GetPFParticles(art::Event const &evt, const std::string &label)
```
I wouldn't indent implementation here. I would add a separator line e.g. `//------`, typically going 140 characters across. And throughout. This really helps break things up, visually.

I personally prefer e.g. `const art::Event &event` to `art::Event const &evt` throughout.

lhwhitehead replied:
I have done this throughout.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraEventUtils.cxx:
```diff
@@ -0,0 +1,150 @@
+        std::vector<art::Ptr<recob::PFParticle>> theseCosmics;
+
+        // We only want primary cosmic rays
+        for (art::Ptr<recob::PFParticle> particle : theseParticles)
```
Pandora style is to prefix pointer variable names with p, e.g. `pParticle` throughout. Although this is larpandora, is already more inhomogeneous in style, and this is only a suggestion.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraEventUtils.cxx:
```diff
@@ -0,0 +1,150 @@
+
+        if (!HasNeutrino(evt,label))
+        {
+          throw cet::exception("LArPandora") << "LArPandoraEventUtils::GetNeutrino --- No neutrino found";
```
Stray whitespace and indentation. Typically use four spaces for each level of indentation.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraEventUtils.h:
```diff
@@ -0,0 +1,133 @@
+ *
+ * @brief Utility containing helpful functions for end users to access products from events
+ *
+ * @author leigh.howard.whitehead@cern.ch
```
We haven't put author names in any of our codebase before, and thought is required about what happens when this class is edited, updated or maintained. You're presumably free to do this if you want (does LArSoft have any rules? - if so, we should follow those rules in larpandora).

lhwhitehead replied:
I think I saw Steve do this in some non-Pandora code so just copied it blindly. I'll delete them!

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraEventUtils.h:
```diff
@@ -0,0 +1,133 @@
+    *
+    * @return vector of art::Ptrs to particles
+    */
+    static const std::vector<art::Ptr<recob::PFParticle>> GetPFParticles(art::Event const &evt, const std::string &label);
```
const by value return is of very little (~zero) benefit and non-const return by value typically preferred from c++11 onwards. I would typically return the filled vector by non-const reference, but this might confuse very new users, and the aim is to be very user friendly.

lhwhitehead replied:
Do you mean return it via a non-const vector argument? I do that internally when using the template functions, but I was worried about the usability implications of using it for the public access functions.

PandoraPFA replied:
Yes, that's right. This const qualifier does nothing for your function implementation, and so you're giving control back to the calling function as to whether they want the vector they receive and own to be const or non-const.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraEventUtils.h:
```diff
@@ -0,0 +1,133 @@
+    static const bool HasNeutrino(art::Event const &evt, const std::string &label);
+    
+
+private:
```
As before.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+namespace lar_pandora
+{
+
+    const std::vector<art::Ptr<anab::T0>> LArPandoraPFParticleUtils::GetT0(const art::Ptr<recob::PFParticle> &part, art::Event const &evt, const std::string &label)
```
Again, I wouldn't indent here.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+
+        std::vector<art::Ptr<recob::PFParticle>> children;
+
+        for (unsigned int p = 0; p < particles->size(); ++p)
```
As per previous discussion. Index with clearer variable name?

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+        std::vector<art::Ptr<recob::Track>> theseTracks;
+        GetAssocProductVector(part,evt,particleLabel,trackLabel,theseTracks);
+
+        if (theseTracks.size() == 0)
```
`if (theseTracks.empty())` is faster (apparently, as doesn't need to actually find out how big a vector may be) and arguably cleaner.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+        {
+            throw cet::exception("LArPandora") << " LArPandoraPFParticleUtils::GetSlice --- No associated slice found";
+        }
+        else{
```
`{` on new line.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+            throw cet::exception("LArPandora") << " LArPandoraPFParticleUtils::GetSlice --- No associated slice found";
+        }
+        else{
+            sliceIndex = static_cast<unsigned int>(metaMap.at("SliceIndex"));
```
Can use iterator cached above here, as per Andy's comment, to save second look up (this will be a small map, though).

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+        // This function needs to fail if GetTrack would fail
+        std::vector<art::Ptr<recob::Track>> theseTracks;
+        GetAssocProductVector(part,evt,particleLabel,trackLabel,theseTracks);
+        if (theseTracks.size() == 0)
```
Or `(!theseTracks.empty())` if that's still clear enough.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.h:
```diff
@@ -0,0 +1,100 @@
+    * 
+    * @return vector of art::Ptrs to the T0(s)
+    */ 
+    static const std::vector<art::Ptr<anab::T0>> GetT0(const art::Ptr<recob::PFParticle> &part, art::Event const &evt, const std::string &label);
```
As per previous discussion about const by value returns.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.h:
```diff
@@ -0,0 +1,100 @@
+    */
+    static const std::vector<art::Ptr<recob::PFParticle>> GetChildParticles(const art::Ptr<recob::PFParticle> &part, art::Event const &evt, const std::string &label);
+
+    static const std::vector<art::Ptr<recob::Hit>> GetHits(const art::Ptr<recob::PFParticle> &part, art::Event const &evt, const std::string &label);
```
All awaiting doxygen details, or deliberately omitted? Lots of const by value returns.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraShowerUtils.cxx:
```diff
@@ -0,0 +1,53 @@
+namespace lar_pandora
+{
+
+    const std::vector<art::Ptr<recob::Hit>> LArPandoraShowerUtils::GetHits(const art::Ptr<recob::Shower> &shower, art::Event const &evt, const std::string &label)
```
I wouldn't indent and would break up function implementations with `//-----` lines.

PandoraPFA commented on larpandora/LArPandoraAnalysisUtils/LArPandoraSliceUtils.cxx:
```diff
@@ -0,0 +1,34 @@
+
+    const std::vector<art::Ptr<recob::Hit>> LArPandoraSliceUtils::GetHits(const art::Ptr<recob::Slice> &slice, art::Event const &evt, const std::string &label)
+    {    
+
```
Stray whitespace plus indenting.

lhwhitehead commented on larpandora/LArPandoraAnalysisUtils/LArPandoraClusterUtils.cxx:
```diff
@@ -0,0 +1,34 @@
+namespace lar_pandora
+{
+
+    const std::vector<art::Ptr<recob::Hit>> LArPandoraClusterUtils::GetHits(const art::Ptr<recob::Cluster> &cluster, art::Event const &evt, const std::string &label)
```
The LArSoft style inside modules is to use `art::Event const &evt`, but I can switch them here.

lhwhitehead commented on larpandora/LArPandoraAnalysisUtils/LArPandoraPFParticleUtils.cxx:
```diff
@@ -0,0 +1,212 @@
+        GetAssocProductVector(part,evt,label,label,theseClusters);
+
+        std::vector<art::Ptr<recob::Hit>> theseHits;
+        for (const art::Ptr<recob::Cluster> cluster : theseClusters)
```
Since the line of code of the for loop isn't long, I'd suggest keeping the variable type for readability purposes.

lhwhitehead commented on larpandora/LArPandoraAnalysisUtils/LArPandoraClusterUtils.cxx:
```diff
@@ -0,0 +1,34 @@
+    {    
+
+        std::vector<art::Ptr<recob::Hit>> theseHits;
+        GetAssocProductVector(cluster,evt,label,label,theseHits);
```
Since the functions are all static I went with the class name prefix.

lhwhitehead commented:
Hi all, I think I've implemented the comments - thanks for all of the suggestions. I think the code is much neater in places now, and I tested that the utility module (now renamed for a dunetpc module name clash reason) gives me the same results as it did before. There are some fairly big changes to the way the templating works, but I think it is now much more maintainable whilst keeping readability for the users.

absolution1 commented on larpandora/LArPandoraAnalysisUtils/LArPandoraUtilsBase.h:
```diff
@@ -0,0 +1,90 @@
+
+    // We need to convert these to art pointers
+    std::vector<art::Ptr<T>> productVector;
+    for(unsigned int productIndex = 0; productIndex < theseProds->size(); ++productIndex)
```
The art library provides a small helper function to hand over an art::Ptr filled vector from a handle

`art::fill_ptr_vector(std::vector<art::Ptr<T> >, art::Handle<std::vector<T> >);`

You could make the template implementations ever so slightly smaller by using it.

PandoraPFA commented:
Hi Leigh - Many thanks for the careful responses! Let's try and chat about this during the next few days, and check we're all on the same page, but (with your testing too), I'd imagine we can get this all over to larsoft very soon. Thanks again!

absolution1 reviewed (approved):
Very nice. 
I don't think I've really found anything that @AndyChappell hasn't already found...
